### PR TITLE
#1738 #1689 #1759 #1751 #1717 #1690 #1739

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
Amounts to the following changes to the formatter and smart indent:
- respect single-line scopes like ```if(TRUE) {1}``` and keep them single line
- indent in function arguments at a regular level rather than align by first argument
- keep user indentation in arguments even if not aligned
- improve smart indent in argument lists with function definitions in them
- respect line breaks and user indent in multi-line arguments such as positioning of )
- update tests and add new ones

Fix VariableExplorer test instability by switching to JSON serializer
Update test baselines reflecting changes

### CODE REVIEW

**Comment by MikhailArkhipov** [comment]
Let me look at couple changes in tests one more time

**Comment by MikhailArkhipov** [comment]
Verified that original file matches so where if looked like

```
if(...) {
}
else ...
```

'else' positioning is respected and not pulled to ```} else```

**Comment by MikhailArkhipov** [comment]
#1771 #1767 #1761 #1113

File: src/Package/Impl/DataInspect/VariableSearchTask.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        protected override void OnStartSearch() {
+            base.OnStartSearch();
+
+            var searchString = SearchQuery.SearchString;
+            if (_grid?.Items != null && !string.IsNullOrWhiteSpace(searchString)) {
+                foreach(var itemControl in _grid.Items) {
+                    var tn = itemControl as ObservableTreeNode;
+                    var model = tn?.Model?.Content as VariableViewModel;
+                    if(model != null) {
+                        if (model.Name.StartsWithOrdinal(searchString)) {
```
Comment (int19h) [inline src/Package/Impl/DataInspect/VariableSearchTask.cs]: I think we should do two passes here, first one using `StartsWith`, and then another one using `Contains` (so that variables that start with it are listed first, but you can also find stuff in the middle).

Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/VariableSearchTask.cs]: It only finds one at the moment. No multiple selection

---
In reply to: [63251922](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63251922) [](ancestors = 63251922)

Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/VariableSearchTask.cs]: I.e. it moves actual selection not highlight

---
In reply to: [63252079](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63252079) [](ancestors = 63252079,63251922)

Comment (int19h) [inline src/Package/Impl/DataInspect/VariableSearchTask.cs]: Ah. Then do the second pass with `Contains` if we didn't find anything with `StartsWith`.

File: src/Package/Impl/DataInspect/VariableWindowPane.cs [hunk 2]
```diff
@@ -21,5 +23,28 @@ internal sealed class VariableWindowPane : RToolWindowPane {
 
             ToolBar = new CommandID(RGuidList.RCmdSetGuid, RPackageCommandId.variableWindowToolBarId);
         }
+
+        public override bool SearchEnabled
```
Comment (int19h) [inline src/Package/Impl/DataInspect/VariableWindowPane.cs]: Formatting.

File: src/Package/Impl/DataInspect/VariableSearchTask.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/VariableSearchTask.cs]: OK

---
In reply to: [63252176](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63252176) [](ancestors = 63252176)

File: src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs [hunk 3]
```diff
@@ -32,7 +33,12 @@ internal abstract class GridViewerBase : ViewerBase, IObjectDetailsViewer {
             if (evaluation != null) {
                 await VsAppShell.Current.SwitchToMainThreadAsync();
 
-                var id = _toolWindowIdBase + evaluation.GetHashCode() % (Int32.MaxValue - _toolWindowIdBase);
+                var id = Math.Abs(_toolWindowIdBase + expression.GetHashCode() % (Int32.MaxValue - _toolWindowIdBase));
+
+                var existingPane = ToolWindowUtilities.FindWindowPane<VariableGridWindowPane>(id);
+                var frame = existingPane?.Frame as IVsWindowFrame;
+                frame?.CloseFrame(0);
```
Comment (int19h) [inline src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs]: If it exists, shouldn't we just focus on it?

File: src/Package/Impl/Help/HelpWindowPane.cs [hunk 4]
```diff
@@ -28,5 +37,64 @@ internal class HelpWindowPane : VisualComponentToolWindow<IHelpVisualComponent>
...
+
+            protected override async void OnStartSearch() {
+                base.OnStartSearch();
+                _callback.ReportComplete(this, 1);
+
+                var searchString = SearchQuery.SearchString;
+                if (!string.IsNullOrWhiteSpace(searchString)) {
+                    var session = _workflowProvider.GetOrCreate().RSession;
+                    using (var inter = await session.BeginInteractionAsync()) {
+                        await inter.RespondAsync("?" + searchString + Environment.NewLine);
```
Comment (int19h) [inline src/Package/Impl/Help/HelpWindowPane.cs]: Use `ExecuteAsync` and `rtvs:::show_help`, like `ShowHelpOnCurrentCommand` (so that it doesn't require an active prompt, and falls back to `??` if `?` didn't work).

File: src/R/Editor/Impl/Completion/Providers/PackageFunctionCompletionProvider.cs [hunk 5]
```diff
@@ -79,6 +80,16 @@ public class PackageFunctionCompletionProvider : IRCompletionListProvider {
         }
         #endregion
 
+        #region IRHelpSearchTermProvider
+        public IReadOnlyCollection<string> GetEntries() {
+            var list = new List<string>();
+            foreach (IPackageInfo pkg in PackageIndex.Packages) {
```
Comment (int19h) [inline src/R/Editor/Impl/Completion/Providers/PackageFunctionCompletionProvider.cs]: Looks like it's `SelectMany`?

File: src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs (hunk 3 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs]: Perhaps. I need to check if it actually updates the data or if the data is detached.

---
In reply to: [63252280](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63252280) [](ancestors = 63252280)

Comment (int19h) [inline src/Package/Impl/DataInspect/Viewers/GridViewerBase.cs]: If it doesn't refresh, I'd just treat it as a separate bug to fix in another PR.

**Comment by int19h** [comment]
:+1:

File: src/Package/Impl/Help/HelpWindowPane.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Help/HelpWindowPane.cs]: It didn't work for me. Test also does Interact. 

---
In reply to: [63252584](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63252584) [](ancestors = 63252584)

Comment (MikhailArkhipov) [inline src/Package/Impl/Help/HelpWindowPane.cs]: OK, will change then test and command to "rtvs:::show_help

---
In reply to: [63253119](https://github.com/Microsoft/RTVS/pull/1765#discussion_r63253119) [](ancestors = 63253119,63252584)

Comment (int19h) [inline src/Package/Impl/Help/HelpWindowPane.cs]: The tricky part with `?` is that it doesn't cause the help to open when it is evaluated. Instead, it returns an object (actually just a string URL with a custom class) that causes help to open when it is printed. So if you just evaluate `?x`, you get that string back, and nothing opens. `show_help` wraps all the stuff that's needed to actually make it open.

On the other hand, if you want to test whether search is successful or not, this means that you can eval `?x` first and check the result. If it didn't find anything, it'll be a 0-length vector. On the other hand, `??` never fails.
